orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -2220,27 +2220,16 @@ Return true if all elements are true and false otherwise.
       .Output(0, "outputs_grad", "Gradient of outputs returned from pytorch.", "T", OpSchema::Variadic,
               /*is_homogeneous*/ false,
               /*min_arity*/ 1)
-      .Attr(
-        "required_grad",
-        "The indices of the outputs that require gradient outputs.",
-        AttributeProto::INTS)
       .TypeConstraint("T", OpSchema::all_tensor_types(), "Allow inputs and outputs to be any kind of tensor.")
       .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-        const std::string attribute_name = "required_grad";
-        auto required_grads = ctx.getAttribute(attribute_name);
-        if (nullptr == required_grads) {  // attribute not present
-          fail_type_inference("Value of attribute ", attribute_name, " not specified");
-        }
-        ORT_ENFORCE(ctx.getNumOutputs() == static_cast<size_t> (required_grads->ints_size()));
-        for (size_t i = 0, n = static_cast<size_t> (required_grads->ints_size()); i < n; ++i) {
-          size_t j = static_cast<size_t> (required_grads->ints(static_cast<int>(i)));
-          ORT_ENFORCE(ctx.getNumInputs() > j);
-          propagateElemTypeFromInputToOutput(ctx, j, i);
-          auto typeProto = ctx.getInputType(j);
+        ORT_ENFORCE(ctx.getNumInputs() == ctx.getNumOutputs());
+        for (size_t i = 0; i < ctx.getNumInputs(); ++i) {
+          propagateElemTypeFromInputToOutput(ctx, i, i);
+          auto typeProto = ctx.getInputType(i);
           if (!hasShape(*typeProto)) {
             continue;
           }
-          propagateShapeFromInputToOutput(ctx, j, i);
+          propagateShapeFromInputToOutput(ctx, i, i);
```

Comment:
shape can be different, right?  for scaler 0 ?

Reply:
Scalar-0 is just a workaround during execution. Logically the shape is same. Setting this here can help to infer other shape info in the backward graph.

Reply:
Aishwarya found an issue in T5 that mem_plan reuse this scalar tensor. I think mem_plan should have some bug, but here I will not set the shape for outputs that might be scalar. Will check the mem_plan issue later.